Provide application level option for controlling new nonzero behavior

lindsayad opened the pull request:
Applications may now override the `MooseApp` default for what to do when
PETSc is forced to allocate a new nonzero entry during
Jacobian/preconditioner construction. The default is to not error when
these allocations happen. The downside to this is is that runtime may be
significantly slowed without the user having any idea that slow-down is
occurring because of new nonzero allocations.

Refs #7901

lindsayad committed:
Provide application level option for controlling new nonzero behavior

Applications may now override the `MooseApp` default for what to do when
PETSc is forced to allocate a new nonzero entry during
Jacobian/preconditioner construction. The default is to not error when
these allocations happen. The downside to this is is that runtime may be
significantly slowed without the user having any idea that slow-down is
occurring because of new nonzero allocations.

Refs #7901

fdkong reviewed (changes requested):
Looks great. Could you add a test in NS module? Check we actually see an error when preallocation is wrong

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/610189/) on 0250db9 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/16047/site)

This comment will be updated on new commits.

lindsayad committed:
Add malloc test to navier-stokes

Actually a really good idea by Fande to add a test because I didn't
realize that the navier-stokes binary uses the `NavierStokesTestApp`
which prior to this commit inherited directly from `MooseApp` and hence
the override of `errorOnJacobianNonzeroReallocation` would have done
nothing

lindsayad commented:
Wow adding a test for this sucks

lindsayad commented:
If this gets run with the combined executable, then it won't work, so pretty damn hard to do this with the test harness

fdkong commented:
> If this gets run with the combined executable, then it won't work, so pretty damn hard to do this with the test harness

I did something like `executable_pattern = 'external_petsc_solver*'` to restrict tests not run with combined-*.

You could do the similar thing 

`executable_pattern = 'navier_stokes*'`

lindsayad commented:
> I did something like `executable_pattern = 'external_petsc_solver*'` to restrict tests not run with combined-*.
> 
> You could do the similar thing
> 
> `executable_pattern = 'navier_stokes*'`

Beautiful. Thank you

moosebuild commented:
Job [External App tests](https://civet.inl.gov/job/610978/) on a7c1b4b : invalidated by @lindsayad

lindsayad commented:
@fdkong this is ready for further review/merge

fdkong commented on modules/navier_stokes/test/src/kernels/MallocKernel.C:
```diff
@@ -0,0 +1,36 @@
+  auto & sys_mat = _sys.getMatrix(_sys.systemMatrixTag());
+
+  // Add to an off-diagional
+  sys_mat.add(0, 1, 0);
```
I feel like `.add(0, N, 0)` would be better. `N` is the global column size of sys_mat. 

I was worried about if we increase mesh, this may not trigger malloc error anymore.

lindsayad replied:
I don't think increasing the mesh size should have any effect here. The only thing that would change this being an error is if we switched the default dof ordering from variable major to node major. Because here I'm using PJFNK with default diagonal coupling. 0 is dof for variable u; 1 is dof for variable v

lindsayad replied:
Is it not a different error if you try to add outside the bounds of the matrix? Wouldn't doing an addition at (0, N) actually be outside the bounds? I would think you can only index up to N - 1.

fdkong replied:
OK, I see. I understood now. I missed that you are using PJFNK.
